sled-agent/src/serial.rs

```diff
@@ -0,0 +1,262 @@
+            .take(max_bytes.unwrap_or(DEFAULT_MAX_LENGTH as usize))
+            .collect();
+        let end_offset = from_start + data.len();
+        Ok((data, end_offset))
+    }
+
+    fn offsets_from_start_and_end(&self, start: isize) -> (usize, usize) {
+        if start >= 0 {
+            if self.total_bytes > start {
+                (start as usize, (self.total_bytes - start) as usize)
+            } else {
+                (self.total_bytes as usize, 0)
+            }
+        } else {
+            if self.total_bytes > -start {
+                ((self.total_bytes + start) as usize, -start as usize)
+            } else {
+                (0, self.total_bytes as usize)
+            }
+        }
+    }
+}
+
+pub(crate) struct SerialConsoleBuffer {
+    task: JoinHandle<()>,
+    data: Arc<RwLock<BufferData>>,
+}
+
+impl SerialConsoleBuffer {
+    pub(crate) fn new(ws_uri: String, log: Logger) -> Self {
+        let data = Arc::new(RwLock::new(BufferData::new(TTY_BUFFER_SIZE)));
+        let data_inner = data.clone();
+        let task = tokio::task::spawn(async move {
+            let connect_future =
+                retry(ExponentialBackoff::default(), || async {
+                    match tokio_tungstenite::connect_async(&ws_uri).await {
+                        Ok(x) => Ok(x),
+                        Err(err) => {
+                            warn!(
+                                log,
+                                "TTY connection to {}: {:?}", &ws_uri, err
+                            );
+                            Err(BackoffError::Transient {
+                                err,
+                                retry_after: None,
+                            })
+                        }
+                    }
+                });
+            match connect_future.await {
+                Ok((mut websocket, _)) => loop {
+                    match websocket.next().await {
+                        None => {
+                            warn!(log, "Nothing read from {}", &ws_uri);
+                        }
+                        Some(Err(e)) => {
+                            error!(
+                                log,
+                                "Reading TTY from {}: {:?}", &ws_uri, e
+                            );
+                        }
+                        Some(Ok(Message::Close(details))) => {
+                            info!(
+                                log,
+                                "Closing TTY connection to {}{}",
+                                ws_uri,
+                                if let Some(cf) = details {
+                                    format!(": {}", cf)
+                                } else {
+                                    String::new()
+                                }
+                            );
+                        }
+                        Some(Ok(msg)) => {
+                            data_inner.write().await.consume(msg.into_data());
```

Comment:
I don't remember if the tungstenite crate handles it transparently, but would this also try to "consume" the WebSocket ping/pong messages?

Reply:
totally would, oops! it *does* handle sending pongs transparently, but i think they're also passed through.